Overhaul documentation contribution guide

stevepiercy opened the pull request:
- [X] Closes #991

## Description

- Improve documentation contribution guide by adding chapters for small edits, builds and checks, and a style guide.
- Added details for Vale usage, Diátaxis framework, narrative and API documentation, and fixing all spelling errors.

## Checklist

- [X] I've added a change log entry to `CHANGES.rst`.
- [X] I've added or updated tests if applicable.
- [X] I've run and ensured all tests pass locally by following [Run tests](https://icalendar.readthedocs.io/en/latest/contribute/development.html#run-tests).
- [X] I've added or edited documentation, both as docstrings to be rendered in the API documentation and narrative documentation, as necessary.



<!-- readthedocs-preview icalendar start -->
----
📚 Documentation preview 📚: 

## Review all pages in the new subsection

https://icalendar--1054.org.readthedocs.build/en/1054/contribute/documentation/index.html

<!-- readthedocs-preview icalendar end -->

niccokunzmann commented:
This is interesting ... let us know what you would like to be checked when you are ready :)

stevepiercy commented:
This is now ready for review. I carved out a new section with a total of four pages.

https://icalendar--1054.org.readthedocs.build/en/1054/contribute/documentation/index.html

niccokunzmann commented:
Hi, thanks. I read through it and this is what came up for me:

- [ ] https://icalendar--1054.org.readthedocs.build/en/1054/contribute/documentation/small-edits.html
  This might require forking. It is good to check that the workflow works with an external, second github account when logged out.
- [ ] Headings  
  "Build and check" and "small edits"
  They could also be "Quickly Edit Documentation Online", "Work on the documentation on your computer"
  These are quite long titles. "Build and check" does not capture **Why I would like to read that section**. How can the why be expressed concisely?
- [ ] https://icalendar--1054.org.readthedocs.build/en/1054/contribute/documentation/build-check.html#continuous-integration-checks
  Could also include the doctest from #1063
- [ ] https://icalendar--1054.org.readthedocs.build/en/1054/contribute/documentation/style-guide.html
  I would prefer to always use **object only** as the preferred method. The fully qualified version is useful only when referring to a method in another class. The `custom text label`, I would discourage, especially with this example. We don't use it and it makes it hard to see what is talked about.
- [ ] https://icalendar--1054.org.readthedocs.build/en/1054/contribute/documentation/style-guide.html#docstring-structure
  has inconsistent formatting for the sub-headings
- https://icalendar--1054.org.readthedocs.build/en/1054/contribute/documentation/style-guide.html#component-register
  The `Example` section is not formatted with `.. code-block:: pycon`
  - [ ] There, we also miss a class and a module docstring. We have function and method.
  - [ ] there are two method docs, I would remove one.
  - [ ] ValueError is raised twice. Is that good?

stevepiercy commented:
1. I changed this accordingly in eb3c1f71a5412403dbb55fd96042dc3dff611eb6.
2. Yeah, many of the docstrings are copy-paste from the RFCs, including indentation and line breaks. These kind are not always well structured. I also omitted a few things and cleaned it up a bit more in 86b85aae24266fcf273140d9641402cf1984f42d.
3. Well, we previously had a "perfect example" that I copied from the Google Python Style Guide, but that meant excluding the entire page from being tested and the code couldn't run anyway. Sure, we could modify the script to exclude that "perfect example", but I don't see the advantage of doing so when it's already published elsewhere. Also, I think it's much more useful to have practical existing examples from icalendar itself that are both tested and integrated with other parts of the documentation. It's not practical to have all the sections in a docstring.

   As far as what to do, see my previous commit, which I think helps clarify this new issue for first-timers. IMO, we should focus on what the _code does_ in the docstrings, and less copy-paste of content from the RFC and more links to explicit sections in the RFC. I just added some content to help with that in 018f85725bfa762a8e229834460686f38336289d.
4. `Arguments` and `Args` are aliases to `Parameters`. I removed `Arguments` because it's not actually the correct term to use for signatures, and I was sloppy. See https://docs.python.org/3/faq/programming.html#what-is-the-difference-between-arguments-and-parameters. There's a link to the full list of supported docstring sections for further reference.
5. Sometimes a one-line summary is all that is necessary for a docstring.

Please have a look at the commits and comments I made, and let me know. Thank you!

niccokunzmann commented:
Cool. The only thing I would change now is the docstrings to use "Parameters" instead of Arguments or Args.

stevepiercy commented:
@niccokunzmann your changes look good. Merging!
